include/boost/iostreams/filter/zlib.hpp

```diff
@@ -347,7 +347,7 @@ bool zlib_compressor_impl<Alloc>::filter
     int result = xdeflate(flush ? zlib::finish : zlib::no_flush);
     after(src_begin, dest_begin, true);
     zlib_error::check BOOST_PREVENT_MACRO_SUBSTITUTION(result);
-    return result != zlib::stream_end; 
+    return result == zlib::okay; 
```

Comment:
It's for handling Z_BUF_ERROR condition.
If you analyze the code carrefully, you can see that the return value of the filter function is used to check the success of the operation, not to check the stream_end.
If we treat Z_BUF_ERROR as an error and throw an exception, then (result != zlib::stream_end) == (result == zlib::okay), because the result may only have two values (okay or stream_end). So in that case the patch is ok.
If we don not treat Z_BUF_ERROR as an error, then we should not return true because the operation is not successfull.